### PR TITLE
[DependencyInjection] [Routing] [Config] Recursive directory loading

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13246, #11045, #11059 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by stof** [comment]
The changelog updates are missing

**Comment by stof** [comment]
@nicolas-grekas could you also have a look at the test failure in the TwigBundle tests ?

File: src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function supports($resource, $type = null)
+    {
+        if ('directory' === $type) {
+            return true;
+        }
+
+        return null === $type && '/' === substr($resource, -1);
```
Comment (ogizanagi) [inline src/Symfony/Component/DependencyInjection/Loader/DirectoryLoader.php]:
```diff
- return null === $type && '/' === substr($resource, -1);
+ return null === $type && is_string($resource) && '/' === substr($resource, -1);
```

will solve the failing test reported by Travis (`$resource` could be a Closure).

**Comment by nicolas-grekas** [comment]
Comments addressed

**Comment by nicolas-grekas** [comment]
ping @symfony/deciders

**Comment by dunglas** [comment]
:+1

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**